---
title: "VDJdb summary statistics"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Make sure to keep the default for normal processing.
default_output_hook <- knitr::knit_hooks$get("output")

# Output hooks handle normal R console output.
knitr::knit_hooks$set( output = function(x, options) {

  comment <- knitr::opts_current$get("comment")
  if( is.na(comment) ) comment <- ""
  can_null <- grepl( paste0( comment, "\\s*\\[\\d?\\]" ),
                     x, perl = TRUE)
  do_null <- isTRUE( knitr::opts_current$get("null_prefix") )
  if( can_null && do_null ) {
    # By default R print output aligns at the right brace.
    align_index <- regexpr( "\\]", x )[1] - 1
    # Two cases: start or newline
    re <- paste0( "^.{", align_index, "}\\]")
    rep <- comment
    x <- gsub( re, rep,  x )
    re <- paste0( "\\\n.{", align_index, "}\\]")
    rep <- paste0( "\n", comment )
    x <- gsub( re, rep,  x )
  }

  default_output_hook( x, options )

})

knitr::opts_template$set("kill_prefix"=list(comment=NA, null_prefix=TRUE))
```

```{r message=FALSE, warning=FALSE}
library(knitr)
library(ggplot2)
library(RColorBrewer)
library(data.table)
library(dplyr)
library(httr)
library(xml2)
library(stringr)
library(gridExtra)
select = dplyr::select

df = fread("../database/vdjdb.slim.txt", header=T, sep="\t")
```

!summary_embed_start!

```{r, null_prefix=TRUE, warning=FALSE}
paste("Last updated on", format(Sys.time(), '%d %B, %Y'))
```

#### Record statistics by species and TCR chain

General statistics. Note that general statistics is computed using the 'slim' database version. This version, for example, lists the same TCR sequence found in several samples/studies only once.

```{r warning=FALSE}
df.sg = df %>% 
  group_by(species, gene) %>%
  summarize(records = length(complex.id), 
            paired.records = sum(ifelse(complex.id=="0",0,1)),
            epitopes = length(unique(antigen.epitope)),
            publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1])))

colnames(df.sg) = c("Species", "Chain", "Records", "Paired records", "Unique epitopes", "Publications")

kable(format = "html", df.sg)
```

#### Summary by MHC alleles

Representative data for Homo Sapiens

```{r warning=FALSE}
df.m = df %>% 
  filter(species == "HomoSapiens") %>%
  mutate(mhc.a.split = str_split_fixed(mhc.a, "[:,]", n = Inf)[,1],
         mhc.b.split = str_split_fixed(mhc.b, "[:,]", n = Inf)[,1]) %>%
  group_by(mhc.a.split, mhc.b.split) %>%
  summarize(records = n(), 
            antigens = length(unique(antigen.epitope)),
            publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1]))) %>%
  arrange(-records)
  
colnames(df.m) = c("First chain", "Second chain", "Records", "Unique epitopes", "Publications")
kable(format = "html", df.m)
```

#### Summary by antigen origin

Representative data for Homo Sapiens

```{r warning=FALSE}
df.a = df %>% 
  filter(species == "HomoSapiens") %>%
  group_by(antigen.species) %>% #, antigen.gene) %>%
  summarize(records = n(), 
            epitopes = length(unique(antigen.epitope)),
            publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1]))) %>%
  arrange(-records)

colnames(df.a) = c("Parent species", #"Parent gene",
                   "Records", "Unique epitopes", "Publications")
kable(format = "html", df.a)
```

#### Distribution VDJdb confidence scores

Legend: 0 - critical information missing, 1 - medium confidence, 2 - high confidence, 3 - very high confidence.

```{r warning=FALSE}
df.score <- df %>%
  group_by(mhc.class, gene, vdjdb.score) %>%
  summarize(total = n())

ggplot(df.score, aes(x=paste(mhc.class, gene, sep = " "), y=total, fill = as.factor(vdjdb.score))) + 
  geom_bar(stat = "identity", color = "black", size = 0.3) +  
  xlab("") + ylab("") +
  scale_fill_brewer("VDJdb score", palette = "PuBuGn") + 
  theme_light() +
  theme(legend.position="bottom")
```

#### Spectratype

Representative data for Homo Sapiens. The spectratype is colored by epitope.

```{r warning=FALSE}
df.spe = subset(df, species=="HomoSapiens")

ggplot(df.spe, 
       aes(x=nchar(as.character(cdr3)), fill = antigen.epitope)) + 
  geom_histogram(bins = 21, size = 0.3) + 
  scale_x_continuous(limits = c(5,25), breaks = seq(5,25,5)) + 
  facet_wrap(~gene, scales="free") + 
  scale_fill_hue(guide=F) +
  xlab("CDR3 length") + ylab("Records") +
  theme_bw()
```

#### Variable segment usage and MHC allele

Representative data for Homo Sapiens, TCR beta chain.

```{r warning=FALSE}
df.vhla = df %>% filter(species == "HomoSapiens" & gene == "TRB") %>%
  mutate(mhc.a.split = str_split_fixed(mhc.a, "[:,]", n = Inf)[,1],
         mhc.b.split = str_split_fixed(mhc.b, "[:,]", n = Inf)[,1],
         v.segm.split = str_split_fixed(v.segm, "[*,]", n = Inf)[,1]) %>%
  group_by(mhc.class, mhc.a.split, mhc.b.split, v.segm.split) %>%
  summarize(records = n()) %>%
  group_by(mhc.class, mhc.a.split, mhc.b.split) %>%
  mutate(records.mhc = sum(records)) %>%
  group_by(v.segm.split) %>%
  mutate(records.v = sum(records))
  
#df.vhla$v.segm.split = with(df.vhla, factor(v.segm.split, v.segm.split[order(records.v)]))
#df.vhla$mhc.a.split = with(df.vhla, factor(mhc.a.split, mhc.a.split[order(records.mhc)]))

ggplot(df.vhla, aes(x=paste(mhc.a.split, mhc.b.split, sep = " / "), y=v.segm.split, fill = pmin(records, 100))) +
  geom_tile() +
  scale_fill_gradientn("Records", colors=colorRampPalette(brewer.pal(9, 'PuBuGn'))(32), 
                       trans="log", breaks = c(1, 10, 100)) +
  xlab("") + ylab("") +
  facet_grid(~mhc.class, scales="free", space="free") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
                       axis.text.y = element_text(size=5),
                     panel.grid.major = element_blank())
```

#### Record statistics by year

```{r message=FALSE, warning=FALSE, echo=FALSE, cache=FALSE}
# Load data 

dt.vdjdb = fread("../database/vdjdb_full.txt", sep = "\t", fill=T)

dt.vdjdb.s = dt.vdjdb %>%
  filter(species != "MacacaMulatta") %>%
  mutate(tcr_key = paste(v.alpha, j.alpha, cdr3.alpha, v.beta, j.beta, cdr3.beta),
         mhc_key = paste(mhc.a, mhc.b),
         chains = ifelse(cdr3.alpha != "", ifelse(cdr3.beta != "", "paired", "TRA"), "TRB")) %>%
  select(reference.id, tcr_key, mhc_key, chains, antigen.epitope, species) %>%
  unique

# Fetch pubmed info

ids = unique((dt.vdjdb.s %>%
  filter(substr(reference.id,1,4)=="PMID") %>%
  mutate(id = str_split_fixed(reference.id, ":", 2)[,2]))$id)

pm_data = content(GET("https://eutils.ncbi.nlm.nih.gov/", 
                      path = "entrez/eutils/esummary.fcgi",
                      query = list(db = "pubmed", 
                                   id = paste0(ids,collapse = ","))))

write_xml(pm_data, "dates.xml")
pm_data_2 = readLines("dates.xml")
tmp=file.remove("dates.xml")

pm_data_id = str_split_fixed(pm_data_2[grepl("<Id>", pm_data_2)], "[<>]", n = Inf)[,3]
pm_data_date = str_split_fixed(pm_data_2[grepl('Name="PubDate"', pm_data_2)], "[<>]", n = Inf)[,3]

dt.pubdate = data.table(
  reference.id = paste0("PMID:", pm_data_id),
  pub_date = str_split_fixed(pm_data_date, " ", n = Inf)[,1]
)

dt.pubdate = rbind(dt.pubdate,
                   data.frame(reference.id = 
                                c("https://www.10xgenomics.com/resources/application-notes/a-new-way-of-exploring-immunity-linking-highly-multiplexed-antigen-recognition-to-immune-repertoire-and-phenotype/#",
                                  "http://mediatum.ub.tum.de/doc/1136748",
                                  "https://github.com/antigenomics/vdjdb-db/issues/288",
                                  "https://github.com/antigenomics/vdjdb-db/issues/193",
                                  "https://github.com/antigenomics/vdjdb-db/issues/252"),
                              pub_date = c(2019, 2014, 2019, 2017, 2017) %>% as.character, stringsAsFactors = F))
```

Number of unique TCR sequences, epitopes, MHC alleles and publications by **publication** year, cumulative plots.

```{r message=FALSE,warning=FALSE}
dt.vdjdb.s2 = dt.vdjdb.s %>%
  merge(dt.pubdate) %>%
  merge(data.table(pub_date2 = unique(dt.pubdate$pub_date)), allow.cartesian = T) %>%
  group_by(pub_date2, chains) %>%
  summarise(tcr_count = length(unique(tcr_key[which(pub_date <= pub_date2)])),
            epi_count = length(unique(antigen.epitope[which(pub_date <= pub_date2)])),
            ref_count = length(unique(reference.id[which(pub_date <= pub_date2)])),
            mhc_count = length(unique(mhc_key[which(pub_date <= pub_date2)])))

p1=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = tcr_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2019, by =2)) +
  scale_color_brewer(palette = "Set1") +
  theme_bw() + ggtitle("Number of unique TCR sequences") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))

p2=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = epi_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2019, by =2)) +
  scale_color_brewer(palette = "Set1") +
  theme_bw() + ggtitle("Number of unique epitopes") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))

p3=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = ref_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2019, by =2)) +
  scale_color_brewer(palette = "Set1") +
  theme_bw() + ggtitle("Number of publications") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))

p4=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = mhc_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2019, by =2)) +
  scale_color_brewer("TCR chain(s)", palette = "Set1") +
  theme_bw() + ggtitle("Number of MHC alleles") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5)) +
  theme(legend.position="bottom")

g_legend = function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
  }

mylegend = g_legend(p4)

grid.arrange(arrangeGrob(p1 + theme(legend.position="none"),
                         p2 + theme(legend.position="none"),
                         p3 + theme(legend.position="none"),
                         p4 + theme(legend.position="none"),
                         nrow=2),
             mylegend, nrow=2,heights=c(10, 1))

fwrite(dt.vdjdb.s2, "vdjdb_stats_pubyear.txt", sep = "\t")
```

!summary_embed_end!

#### Unique CDR3 sequences per epitope

Representative data for Homo Sapiens

```{r warning=FALSE}
df.ag = df %>%
  filter(species == "HomoSapiens") %>%
  group_by(gene, mhc.class, antigen.species, antigen.gene, antigen.epitope) %>%
  summarize(cdr3.count = length(unique(cdr3))) %>%
  arrange(-cdr3.count)

ggplot(df.ag, aes(x = cdr3.count)) +
  geom_histogram() +
  scale_x_log10("Unique CDR3s") + ylab("Number of epitopes") +
  facet_wrap(~gene) +
  theme_bw()

colnames(df.ag) = c("TCR chain", "MHC class", "Parent species", "Parent gene", "Epitope sequence", "Unique CDR3")
kable(format = "html", df.ag)
```

Epitopes, parent genes and species

```{r}
df %>%
  filter(species == "HomoSapiens") %>%
  select(antigen.species, antigen.gene, antigen.epitope) %>%
  unique %>%
  group_by(antigen.epitope) %>%
  mutate(ambiguous.naming = ifelse(n() > 1, "*", "")) %>%
  arrange(antigen.species, antigen.gene, antigen.epitope) %>%
  kable(format = "html")
```

```{r}
#
```